Update to support stable PyO3

omerbenamram opened the pull request:
The code works also with PyO3 0.12 + numpy 0.12, but not with PyO3 0.13.
The tests however require PyO3 0.11 + numpy 0.11 + nightly - because of `inline-python` dependency.

All tests are passing :)

This was formatted using latest `rustfmt`, so if there's any styling that needs to be applied we should try and add it to `rustfmt.toml`.

Let me know if there's anything else that needs to be addressed to merge this.

Commit:
Update to support PyO3 0.11

Commit:
fixed a test

de-vri-es commented:
Hey, thank you for the PR. Updating the dependencies has been on the todo-list for a while.

It's very hard to review now because of all the style changes though. You're right that we should include custom formatting in `rustfmt.toml`, but it shouldn't be done in this PR. Do you still have a version without the formatting changes?

de-vri-es commented:
I've pushed a rustfmt.toml and .editorconfig to master, along with applying the formatting to the crate. I've also updated your commits based on that, but you appear to have unticked the box to allow me to push to your branch. So if you want to can update this PR using the `update-to-stable` branch from this repo, or you can manually update your branch.

omerbenamram committed:
Merge editorconfig

omerbenamram committed:
reformat code

omerbenamram commented:
@de-vri-es Thanks!, done :) only the actual changes will be visible now. 

BTW I didn't know there's an option to allow upstream to push to my branch, I'll look for that.

de-vri-es commented on src/from_numpy.rs:
```diff
@@ -57,17 +57,16 @@ pub struct UnalignedArrayError;
+pub unsafe fn matrix_slice_from_numpy<'a, N, R, C>(
 	_py: pyo3::Python,
-	input: O,
+	input: &'a PyAny,
```
Why did you change this argument? It seems like an unnecessary backwards incompatibility, and less convenient for the caller.

omerbenamram replied:
https://docs.rs/pyo3/0.11.1/pyo3/types/struct.PyDict.html#method.get_item now returns `&'a PyAny.` 
And `PyAny ` no longer implements `AsRef<PyAny>`, so this bound prevents the code from compiling.

de-vri-es replied:
Ah, that is a shame. But fair enough, then this seems best for now.

de-vri-es commented on src/from_numpy.rs:
```diff
@@ -203,8 +200,9 @@ where
 		return Err(make_error());
 	}
 
+	let t2: u32 = N::DATA_TYPE.into_ffi_dtype() as u32;
```
```suggestion
```

de-vri-es commented on src/from_numpy.rs:
```diff
@@ -203,8 +200,9 @@ where
+	let t2: u32 = N::DATA_TYPE.into_ffi_dtype() as u32;
 	// Check the data type of the input array.
-	if npyffi::array::PY_ARRAY_API.PyArray_EquivTypenums((*(*array).descr).type_num, N::typenum_default()) != 1 {
+	if npyffi::array::PY_ARRAY_API.PyArray_EquivTypenums((*(*array).descr).type_num, t2 as i32) != 1 {
```
```suggestion
	if npyffi::array::PY_ARRAY_API.PyArray_EquivTypenums((*(*array).descr).type_num, N::ffi_dtype() as u32 as i32) != 1 {
```

de-vri-es commented on src/to_numpy.rs:
```diff
@@ -17,7 +17,7 @@ where
 		for c in 0..matrix.ncols() {
 			unsafe {
-				*array.uget_mut((r, c)) = matrix[(r, c)];
+				*array.uget_mut((r, c)) = matrix[(r, c)].clone();
```
I pretty sure this is not needed, since `nalgebra::Scalar` implies `Copy`.

omerbenamram replied:
![image](https://user-images.githubusercontent.com/2467993/106932213-21e1e980-6720-11eb-981b-51f4ad63fa4e.png)

But AFAIK clone is specifically optimized for scalars - https://docs.rs/nalgebra/0.24.1/src/nalgebra/base/scalar.rs.html#26-28

de-vri-es replied:
Ah, they dropped the `Copy` from `Scalar`. Fair enough. The trait documentation seems to suggest we should use `inlinded_clone()` to opt in to the inlined version in debug builds too. Could you still update that?

de-vri-es replied:
Actually, let's just merge. I'll push a separate commit for that :)

omerbenamram replied:
done.

de-vri-es replied:
Ah, you beat me to it. Even better ;)

de-vri-es commented on tests/from_numpy.rs:
```diff
@@ -11,201 +11,159 @@ mod assert;
 
 /// Test conversion of a numpy array to a Matrix3<f64>.
 #[test]
-#[rustfmt::skip]
```
I added these on purpose to prevent the matrices from being squashed to a single line. I'm guessing this is the result of automatic merge conflict resolving, but could you restore the formatting of the matrices in the tests?

omerbenamram replied:
Whoops, deficiently merge resolving. Reverted and restored matricies.

de-vri-es commented on Cargo.toml:
```diff
@@ -13,10 +13,10 @@ keywords = ["python", "numpy", "nalgebra", "matrix", "conversion"]
-pyo3 = "0.8.2"
+nalgebra = "0.24.1"
+numpy = "0.11"
+pyo3 = { version = "0.11" }
```
```suggestion
pyo3 = "0.11" 
```

de-vri-es commented on src/from_numpy.rs:
```diff
@@ -1,7 +1,7 @@
 use nalgebra::{Dynamic, Matrix};
 use numpy::npyffi;
-use numpy::npyffi::objects::PyArrayObject;
+use numpy::{npyffi::objects::PyArrayObject, Element};
```
It looks like you're using `numpy::Element` throughout the file. I like the explicitness of that, but I think that makes this an unused import.

```suggestion
use numpy::{npyffi::objects::PyArrayObject};
```

de-vri-es reviewed (commented):
Thanks for updating the PR. I left a few in-line comments, but overall it looks good. Thanks also for updating the tests to the  `inline-python` API!

Sadly there are still issues that prevent `inline-python` from updating to `pyo3` 0.13, so we're kinda stuck until that can be resolved.

omerbenamram committed:
Update Cargo.toml

Co-authored-by: Maarten de Vries <maarten@de-vri.es>

omerbenamram committed:
Update src/from_numpy.rs

Co-authored-by: Maarten de Vries <maarten@de-vri.es>

omerbenamram committed:
reapply some lost formatting

omerbenamram committed:
Update src/from_numpy.rs

Co-authored-by: Maarten de Vries <maarten@de-vri.es>

omerbenamram committed:
Update src/from_numpy.rs

Co-authored-by: Maarten de Vries <maarten@de-vri.es>

omerbenamram commented:
Thanks for the thorough CR. I've addressed the comments.

The reason I've settled with 0.11 is mostly because it's the first version to fully support stable rust.
I've looked into the 0.13 issues, and it appears there is indeed work to be done to support it both in the main code and in the tests, but IMO it could be done separately from this change set.

WDYT?

omerbenamram committed:
use `inlined_clone` explicitly

de-vri-es commented:
> Thanks for the thorough CR. I've addressed the comments.

Thank you!
 
> The reason I've settled with 0.11 is mostly because it's the first version to fully support stable rust.
> I've looked into the 0.13 issues, and it appears there is indeed work to be done to support it both in the main code and in the tests, but IMO it could be done separately from this change set.

Yeah, that's good for now. It's already a big improvement :)

> WDYT?

Ready to merge :) I'll try to get a release out today too.

Thanks again for picking this up!

de-vri-es commented:
Released as 0.3.0 :)
